### ISSUE TITLE
The val_acc with validation_steps=validation_steps // hvd.size() is wrong.

### ISSUE THREAD

**Issue by Bonsen** [horovod/horovod#2897]
I run:
```
def masked_categorical_accuracy(y_true, y_pred): (with ignored labels)
    mask = K.not_equal(tf.reduce_max(y_true, axis=1), 0)
    y_true_masked = tf.boolean_mask(y_true, mask)
    y_pred_masked = tf.boolean_mask(y_pred, mask)
    return keras.metrics.categorical_accuracy(y_true_masked, y_pred_masked)

 history = model.fit(
        train_data,
        steps_per_epoch=train_steps // hvd.size(),
        epochs=epochs,
        validation_data=validation_data,
        validation_steps=validation_steps // hvd.size(),
        callbacks=callbacks,
        shuffle=False,
        verbose=verbose)
.....
callback = hvd.callbacks.MetricAverageCallback()
.....
```
The val_acc is wrong. If I run with validation_steps=validation_steps , the val_acc is right. 

------

So I think it's caused by ignored labels that each gpu has different size of batch.

Then I run:
```
def masked_categorical_accuracy(y_true, y_pred): (with ignored labels)
    mask = K.not_equal(tf.reduce_max(y_true, axis=1), 0)
    y_true_masked = tf.boolean_mask(y_true, mask)
    y_pred_masked = tf.boolean_mask(y_pred, mask)
    target = tf.argmax(y_true_masked, axis = 1)
    output = tf.argmax(y_pred_masked, axis = 1)
    rt = tf.reduce_sum(tf.cast(output==target, dtype=tf.float32))
    wg = tf.reduce_sum(tf.cast(output!=target, dtype=tf.float32))
    rt = horovod.tensorflow.allreduce(rt, op=horovod.tensorflow.mpi_ops.Sum)
    wg = horovod.tensorflow.allreduce(wg, op=horovod.tensorflow.mpi_ops.Sum)
    acc = rt / (rt + wg)
    return acc

 history = model.fit(
        train_data,
        steps_per_epoch=train_steps // hvd.size(),
        epochs=epochs,
        validation_data=validation_data,
        validation_steps=validation_steps // hvd.size(),
        callbacks=callbacks,
        shuffle=False,
        verbose=verbose)
```
The new masked_categorical_accuracy collect all right_pred and wrong_pred to get right acc though all gpus. But the result is still wrong.
